File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/08．机制触发/03．单位数量状态监听.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;

const IsUnitType = jass.IsUnitType as (whichUnit: any, whichUnitType: any) => boolean;
const UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD as any;

const { addPeriodicCallback, removePeriodicCallback } = require("系统.00．核心系统.05．中心计时器") as {
  addPeriodicCallback: (this: void, intervalMs: number, callback: (this: void) => void) => number;
  removePeriodicCallback: (this: void, id: number) => void;
};

export interface 单位数量状态清理篮子 {
  登记清理(名称: string, 清理: (this: void) => void): void;
  登记周期回调?(名称: string, 回调ID: number): void;
}

export interface 单位数量状态监听参数 {
  名称?: string;
  单位列表?: any[];
  读取单位列表?: (this: void) => any[];
  监听间隔毫秒?: number;
  清理篮子?: 单位数量状态清理篮子;
  过滤单位?: (this: void, 单位: any) => boolean;
  on数量变化?: (this: void, 当前数量: number, 上次数量: number) => void;
  on刷新?: (this: void, 当前数量: number) => void;
}

export interface 单位数量状态监听控制器 {
  readonly 名称: string;
  刷新(): number;
  读取当前数量(): number;
  停止(): void;
}

function 单位有效(this: void, 单位: any): boolean {
  return 单位 != null && 单位 !== 0 && IsUnitType(单位, UNIT_TYPE_DEAD) !== true;
}

function 读取列表(this: void, 参数: 单位数量状态监听参数): any[] {
  if (参数.读取单位列表 != null) return 参数.读取单位列表();
  return 参数.单位列表 ?? [];
}

function 统计数量(this: void, 参数: 单位数量状态监听参数): number {
  const 列表 = 读取列表(参数);
  let 数量 = 0;
  for (let i = 0; i < 列表.length; i++) {
    const 单位 = 列表[i];
    if (!单位有效(单位)) continue;
    if (参数.过滤单位 != null && !参数.过滤单位(单位)) continue;
    数量 += 1;
  }
  return 数量;
}

class 单位数量状态监听控制器实现 implements 单位数量状态监听控制器 {
  readonly 名称: string;
  private 参数: 单位数量状态监听参数;
  private 当前数量 = -1;
  private 周期回调ID = 0;
  private 已停止 = false;

  constructor(名称: string, 参数: 单位数量状态监听参数) {
    this.名称 = 名称;
    this.参数 = 参数;
  }

  设置周期回调ID(id: number): void {
    this.周期回调ID = id;
  }

  刷新(): number {
    if (this.已停止) return this.当前数量;
    const 上次数量 = this.当前数量;
    const 当前数量 = 统计数量(this.参数);
    this.当前数量 = 当前数量;
    if (this.参数.on刷新 != null) this.参数.on刷新(当前数量);
    if (上次数量 !== -1 && 当前数量 !== 上次数量 && this.参数.on数量变化 != null) {
      this.参数.on数量变化(当前数量, 上次数量);
    }
    return 当前数量;
  }

  读取当前数量(): number {
    return this.当前数量 < 0 ? 0 : this.当前数量;
  }

  停止(): void {
    if (this.已停止) return;
    this.已停止 = true;
    if (this.周期回调ID !== 0) {
      removePeriodicCallback(this.周期回调ID);
      this.周期回调ID = 0;
    }
  }
}

export function 创建单位数量状态监听(this: void, 参数: 单位数量状态监听参数): 单位数量状态监听控制器 {
  const 名称 = 参数.名称 ?? "单位数量状态监听";
  const 控制器 = new 单位数量状态监听控制器实现(名称, 参数);
  控制器.刷新();
  const 间隔 = 参数.监听间隔毫秒 ?? 500;
  if (间隔 > 0) {
    const id = addPeriodicCallback(间隔, function 单位数量状态监听Tick(this: void): void {
      控制器.刷新();
    });
    控制器.设置周期回调ID(id);
    if (参数.清理篮子 != null) {
      if (参数.清理篮子.登记周期回调 != null) 参数.清理篮子.登记周期回调(`${名称}-周期刷新`, id);
      else 参数.清理篮子.登记清理(`${名称}-停止`, function 停止单位数量状态监听(this: void): void {
        控制器.停止();
      });
    }
  }
  return 控制器;
}

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/08．机制触发/03．单位数量状态监听.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local IsUnitType = jass.IsUnitType
local UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local addPeriodicCallback = ____require_result_0.addPeriodicCallback
local removePeriodicCallback = ____require_result_0.removePeriodicCallback
local function _____5355_4F4D_6709_6548(_____5355_4F4D)
    return _____5355_4F4D ~= nil and _____5355_4F4D ~= 0 and IsUnitType(_____5355_4F4D, UNIT_TYPE_DEAD) ~= true
end
local function _____8BFB_53D6_5217_8868(_____53C2_6570)
    if _____53C2_6570["读取单位列表"] ~= nil then
        return _____53C2_6570["读取单位列表"]()
    end
    return _____53C2_6570["单位列表"] or ({})
end
local function _____7EDF_8BA1_6570_91CF(_____53C2_6570)
    local _____5217_8868 = _____8BFB_53D6_5217_8868(_____53C2_6570)
    local _____6570_91CF = 0
    do
        local i = 0
        while i < #_____5217_8868 do
            do
                local _____5355_4F4D = _____5217_8868[i + 1]
                if not _____5355_4F4D_6709_6548(_____5355_4F4D) then
                    goto __continue7
                end
                if _____53C2_6570["过滤单位"] ~= nil and not _____53C2_6570["过滤单位"](_____5355_4F4D) then
                    goto __continue7
                end
                _____6570_91CF = _____6570_91CF + 1
            end
            ::__continue7::
            i = i + 1
        end
    end
    return _____6570_91CF
end
local _____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0 = __TS__Class()
_____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0.name = "单位数量状态监听控制器实现"
function _____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0.prototype.____constructor(self, _____540D_79F0, _____53C2_6570)
    self["当前数量"] = -1
    self["周期回调ID"] = 0
    self["已停止"] = false
    self["名称"] = _____540D_79F0
    self["参数"] = _____53C2_6570
end
_____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0.prototype["设置周期回调ID"] = function(self, id)
    self["周期回调ID"] = id
end
_____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0.prototype["刷新"] = function(self)
    if self["已停止"] then
        return self["当前数量"]
    end
    local _____4E0A_6B21_6570_91CF = self["当前数量"]
    local _____5F53_524D_6570_91CF = _____7EDF_8BA1_6570_91CF(self["参数"])
    self["当前数量"] = _____5F53_524D_6570_91CF
    if self["参数"]["on刷新"] ~= nil then
        self["参数"]["on刷新"](_____5F53_524D_6570_91CF)
    end
    if _____4E0A_6B21_6570_91CF ~= -1 and _____5F53_524D_6570_91CF ~= _____4E0A_6B21_6570_91CF and self["参数"]["on数量变化"] ~= nil then
        self["参数"]["on数量变化"](_____5F53_524D_6570_91CF, _____4E0A_6B21_6570_91CF)
    end
    return _____5F53_524D_6570_91CF
end
_____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0.prototype["读取当前数量"] = function(self)
    return self["当前数量"] < 0 and 0 or self["当前数量"]
end
_____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0.prototype["停止"] = function(self)
    if self["已停止"] then
        return
    end
    self["已停止"] = true
    if self["周期回调ID"] ~= 0 then
        removePeriodicCallback(self["周期回调ID"])
        self["周期回调ID"] = 0
    end
end
____exports["创建单位数量状态监听"] = function(_____53C2_6570)
    local _____540D_79F0 = _____53C2_6570["名称"] or "单位数量状态监听"
    local _____63A7_5236_5668 = __TS__New(_____5355_4F4D_6570_91CF_72B6_6001_76D1_542C_63A7_5236_5668_5B9E_73B0, _____540D_79F0, _____53C2_6570)
    _____63A7_5236_5668["刷新"](_____63A7_5236_5668)
    local _____95F4_9694 = _____53C2_6570["监听间隔毫秒"] or 500
    if _____95F4_9694 > 0 then
        local id = addPeriodicCallback(
            _____95F4_9694,
            function()
                _____63A7_5236_5668["刷新"](_____63A7_5236_5668)
            end
        )
        _____63A7_5236_5668["设置周期回调ID"](_____63A7_5236_5668, id)
        if _____53C2_6570["清理篮子"] ~= nil then
            if _____53C2_6570["清理篮子"]["登记周期回调"] ~= nil then
                local ____self_1 = _____53C2_6570["清理篮子"]
                ____self_1["登记周期回调"](____self_1, _____540D_79F0 .. "-周期刷新", id)
            else
                local ____self_2 = _____53C2_6570["清理篮子"]
                ____self_2["登记清理"](
                    ____self_2,
                    _____540D_79F0 .. "-停止",
                    function()
                        _____63A7_5236_5668["停止"](_____63A7_5236_5668)
                    end
                )
            end
        end
    end
    return _____63A7_5236_5668
end
return ____exports
